Update typescript to 4.4.4 and add new Imgix params

nim442 opened the pull request:
This PR includes two main changes
1. Updates TS to 4.4.4 which allows us to use some shiny new TS features and reduce some boilerplate
2. Adds new parameters some of which are base64 variant. Imgix supports passing the values as base64 as long as we append 64 to the parameter name. See their [api doc](https://docs.imgix.com/apis/rendering#base64-variants) for more info. (Basically, `?txt=Hello%2C%20World!` is equivalent to `?txt64=SGVsbG8sIFdvcmxkIQ`).
In this PR, the base64 variants are typed as string and we assume that it's the library's consumer's job to encode these values. Originally I had thought of encoding the value in the library instead user but had some concerns detailed below:
    - If it is the library that does the encoding, we need to make it SSR friendly somehow. My gut feel is that this should be handled in the userland since the consumer will know when the code is executing on a node js server or a browser. Imgix's own client [library](https://github.com/imgix/js-core/blob/main/src/index.js#L76..L79) base64 encodes the values for the user. Note that they are using an npm package that rolls out it's own platform agnostic base64 function 

    - Alternatively, if it is the consumer who encodes the value, we need to make sure they don't pass in an unencoded string. [Here](https://github.com/unsplash/ts-imgix/commit/cf7d8908a974bbc71d7d0d35a542964917128195) is a quick prototype on what that might look like. In this prototype, a base64 parameter is a branded type that can only be constructed via a constructor function. The constructor function takes an encoder so that the end user can pass their own encoding function which is different depending on if the code is being execute in node.js or browser. This ensures that you can't simply pass an unencoded string.

Commit:
update typescript to 4.4.4 and prettier to 2.4.1

for some nice new ts features

Commit:
rename pipe to flow and use the updated version from fp-ts

pipe function is outdated. This version of pipe is now called flow in fp-ts

OliverJAsh committed:
Add `fm`

Fixes https://github.com/unsplash/ts-imgix/issues/16

OliverJAsh committed:
v0.0.16

Commit:
add new parameters and use the new string literal types to automatically convert paramater names from kebab to camel case

Commit:
add more params

Commit:
remove unused Omit

Commit:
use typed ImgixBlendMode enum instead of a string

nim442 commented on src/helpers/flow.ts:
```diff
@@ -0,0 +1,116 @@
+ * @function
+ * @since 2.11.5
+ */
+export function flow<A extends ReadonlyArray<unknown>, B>(ab: (...a: A) => B): (...a: A) => B;
```
Originally, this was called pipe and was copied from an older version of fp-ts where pipe was defined as flow.

nim442 commented on src/index.ts:
```diff
@@ -58,6 +90,11 @@ export type ImgixRect = {
+// https://docs.imgix.com/apis/rendering/watermark/mark-align
+export type ImgixMarkAlignBase = 'top' | 'middle' | 'bottom' | 'left' | 'center' | 'right';
+// The api allows the user to combine the align values with a comma
+export type ImgixMarkAlign = `${ImgixMarkAlignBase},${ImgixMarkAlignBase}` | ImgixMarkAlignBase;
```
The api allows the user to combine two values with comma like `top,middle`. One way of handling this is to make a Record<ImgixMarkAlignbase,boolean> or a tuple. But this is so much better UX for the end user imo. TS lsp will show a nice dropdown with fuzzy matching for all the possible values.

![image](https://user-images.githubusercontent.com/7310201/141251202-abea466f-daea-4922-80df-c3ee6a13bb8e.png)

Magellol replied:
I guess this notation would also allow `right,right` or `middle,middle` which I think isn't valid values? since we'd be targeting the same axis? One way to avoid that would just to be typing all possible values as part of the union? 

```ts
export type Markalign = "top-left" | "top-right" | ...
```

I think this way is a bit more type safe. WDYT?
Also it looks like `unset` is another possible value but perhaps we don't need it.

nim442 replied:
right. This is the kind of thing where i'd spend hours trying to figure it out a generic way to remove duplicated with TS instead of manually writing it, only to end up realizing that manually typing would have been so much faster and easier

nim442 replied:
it works with this https://github.com/unsplash/ts-imgix/pull/17/commits/5d090cd46a3a4caeede646aa8ea7494f183534d1

Commit:
add txt-pad param

Commit:
add text-width param

Commit:
add txt-clip param

Magellol commented on src/index.ts:
```diff
@@ -76,54 +120,103 @@ export type ImgixUrlQueryParams = {
+    'mark-w'?: number;
+    'mark-align'?: ImgixMarkAlign;
+    'mark-pad'?: number;
+    markY?: number;
```
Shouldn't this be `mark-y`? Do they support both notations?

nim442 replied:
good catch! I think they do support it but it's not documented

OliverJAsh commented on src/index.ts:
```diff
@@ -76,54 +120,103 @@ export type ImgixUrlQueryParams = {
-export type QueryParamsInput = Omit<ImgixUrlQueryParams, 'min-h'> & { minH?: number };
+type KebabToCamelCase<S extends string> = S extends `${infer T}-${infer U}`
+    ? `${T}${Capitalize<KebabToCamelCase<U>>}`
+    : S;
```
Nice!

OliverJAsh reviewed (approved):
When you need to release to npm I highly recommend `np` to make sure you don't miss any of the steps.

Commit:
fix wrong case in some params

Commit:
exclude repeating combination in ImgixMarkAlign

we don't want "left,left" as one of the options for example as they are invalid values.

Commit:
update rollup-plugin-typescript2 and add tslib

tslibi needs to be installed because of this error: https://github.com/alexjoverm/typescript-library-starter/issues/310

Commit:
Merge branch 'master' of https://github.com/unsplash/ts-imgix into update-props
